src/Hamiltonians/HOCartesianEnergyConserved.jl

```diff
@@ -0,0 +1,399 @@
+end
+largest_two_point_interval(i, j, w) = largest_two_point_interval(i, j, 1, w)
+
+function two_point_interval_bounds(i, j, v, w)
+    if i ≤ j
+        left_gap = min(i - v, w - j)
+        left_edge = i - left_gap
+        right_gap = min(w - i, j - v)
+        right_edge = i + right_gap
+    else
+        left_gap = min(j - v, w - i)
+        left_edge = j - left_gap
+        right_gap = min(w - j, i - v)
+        right_edge = j + right_gap
+    end
+    return left_edge, right_edge
+end
+
+# useful conversion of aspect ratio to a factor for getting box size
+aspect_to_box(η::NTuple{N,Int}) where {N} = Int.(lcm(η...) .// η)
+box_to_aspect(S) = (S[1] - 1) ./ (S .- 1)
+box_to_aspect(S::NTuple{N,Int}) where {N} = (S[1] - 1) .// (S .- 1)
+
+@inline function mode_to_energy(i, S, aspect)
+    # aspect = box_to_aspect(S)
+    states = CartesianIndices(S)
+    # account for 1-indexing here in case aspect ratio is fractional
+    idx = Tuple(states[i]) .- 1
+    E = dot(idx, aspect)
+    # shift back to 1-indexing
+    return E + 1
+end
+
+"""
+    find_Ebounds(i, j, S)
+
+Find the range of single particle energies that a particle in mode `i` can move to while 
+preserving total energy of particles `i` and `j`, and constrained to box `S`.
+"""
+function find_Ebounds(i, j, S, aspect)
+    
+    E_i = mode_to_energy(i, S, aspect)
+    E_j = mode_to_energy(j, S, aspect)
+
+    # shift back to 1-indexing
+    return extrema(largest_two_point_interval(E_i, E_j, S[1]))..., E_i + E_j
+end
+
+"""
+    OccupiedPairsMap(omm::OccupiedModeMap) <: AbstractVector
+
+Get all pairs of particles in `omm`. Pairs involving multiply-occupied modes 
+are counted once, (including self pairing).
+This is an eager iterator whose elements are a tuple of particle indices.
+
+See [`OccupiedModeMap`](@ref).
+"""
+struct OccupiedPairsMap{N,T} <: AbstractVector{T}
+    pairs::SVector{N,T}
+    length::Int
+end
+
+function OccupiedPairsMap(addr::SingleComponentFockAddress{N}) where {N}
+    omm = OccupiedModeMap(addr)
+    T = eltype(omm)
+    P = N * (N - 1) ÷ 2
+    pairs = MVector{P,Tuple{T,T}}(undef)
+    L = 0
+    for i in eachindex(omm)
+        p_i = omm[i]
+        if p_i.occnum > 1
+            L += 1
+            @inbounds pairs[L] = (p_i, p_i)
+        end
+        for j in 1:i-1
+            p_j = omm[j]
+            L += 1
+            @inbounds pairs[L] = (p_i, p_j)
+        end
+    end
+    
+    return OccupiedPairsMap(SVector(pairs), L)
+end
+
+Base.size(opm::OccupiedPairsMap) = (opm.length,)
+function Base.getindex(opm::OccupiedPairsMap, i)
+    @boundscheck 1 ≤ i ≤ opm.length || throw(BoundsError(opm, i))
+    return opm.pairs[i]
+end
+
+"""
+    HOCartesianEnergyConserved(addr; S, η, g = 1.0, interaction_only = false)
+
+Implements a harmonic oscillator in Cartesian basis with contact interactions.
+```math
+\\hat{H} = \\sum_{i} \\epsilon_i n_i + \\frac{g}{2}\\sum_{ijkl} V_{ijkl} a^†_i a^†_j a_k a_l
+```
```

Comment:
```suggestion
Implements a harmonic oscillator in Cartesian basis with contact interactions 
in first-order degenerate perturbation theory.
```math
\\hat{H} = \\sum_{i} \\epsilon_i n_i + \\frac{g}{2}\\sum_{ijkl} V_{ijkl} a^†_i a^†_j a_k a_l\\, 
\\delta_{\\epsilon_i + \\epsilon_j}^{\\epsilon_k +\\epsilon_l}
```
```